newm-server/src/main/kotlin/io/newm/server/features/arweave/repo/ArweaveRepositoryImpl.kt

```diff
@@ -0,0 +1,265 @@
+    }
+    private val executorService by lazy { Executors.newScheduledThreadPool(5) }
+    private val executionContext by lazy { ExecutionContext.fromExecutorService(executorService) }
+    private val arweaveBackend by lazy {
+        AsyncHttpClientFutureBackend.apply(
+            SttpBackendOptions.Default(),
+            executionContext
+        )
+    }
+    private val arweaveConfig by lazy { configApi.apply(arweaveUri, arweaveBackend) }
+
+    private suspend fun arweaveWallet(): Wallet {
+        val walletJson = environment.getSecureConfigString("arweave.walletJson")
+        return walletApi.load(Source.fromString(walletJson)).get()
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private suspend fun calculatePrice(data: Data): Winston {
+        val priceFuture = priceApi.dataTransaction(
+            data,
+            arweaveConfig,
+            future.futureJsonHandlerEncodedStringHandler(executionContext)
+        ) as Future<Winston>
+
+        return toJava(priceFuture).await()
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private suspend fun getLastWalletTransaction(): Option<Transaction.Id> {
+        val lastTxFuture = addressApi.lastTx(
+            arweaveWallet().address(),
+            arweaveConfig,
+            future.futureJsonHandlerEncodedStringHandler(executionContext)
+        ) as Future<Option<Transaction.Id>>
+        return toJava(lastTxFuture).await()
+    }
+
+    private fun buildTags(tagMap: Map<String, String>): Option<Seq<Tag.Custom>> {
+        return Option.apply(
+            JavaConverters.collectionAsScalaIterableConverter(
+                tagMap.map { (key, value) ->
+                    tagApi.apply(key.toByteArray(), value.toByteArray())
+                }
+            ).asScala().toSeq()
+        )
+    }
+
+    private suspend fun signTransaction(transaction: Transaction): Signed<Transaction> {
+        return signableApi.SignableSyntax(transaction).sign(arweaveWallet().priv())
+    }
+
+    private suspend fun submitTransaction(signedTransaction: Signed<Transaction>): Boolean {
+        val submitFuture = txApi.submit(signedTransaction, arweaveConfig, handlerFunction)
+        val submitResponse = toJava(submitFuture).await() as Response<*>
+        when (submitResponse.code()) {
+            200 -> {
+                /* no-op */
+            }
+
+            208 -> log.warn("Transaction already processed: $signedTransaction")
+            429 -> log.error("Too many requests!")
+            400, 503 -> log.error("Transaction verification failed!: $signedTransaction")
+            else -> log.error("Unknown problem submitting transaction: $submitResponse")
+        }
+        return submitResponse.isSuccess
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    private suspend fun awaitTransactionSettlement(transactionId: String) {
+        val txId = Transaction.Id(transactionId.b64ToByteArray())
+        withTimeout(Duration.ofHours(3L)) {
```

Comment:
Should duration be a .conf param?

Reply:
This is hardcoded to the same value that we set on our SQS queue. Arweave transactions can take up to 150 minutes (2.5 hours) to settle, so this allows us to continue checking for 3 hours to see if we made it onto the arweave chain.